### PR TITLE
Convert AISupportPowerManager to module

### PR DESCRIPTION
**Pull request by reaperrr**
Follow-up to #15750.

This already was a semi-module, so the changes are comparatively small (the diff of the module looks larger than it really is, because the Info part shifted the rest of the code downwards).

Yaml changes were done via update rule, but then redundancies removed (shipping mods used identical settings across all AIs).

### CODE REVIEW

File: OpenRA.Mods.Common/OpenRA.Mods.Common.csproj [hunk 1]
```diff
@@ -126,15 +126,15 @@
     <Compile Include="AI\BotOrderManager.cs" />
     <Compile Include="AI\HackyAI.cs" />
     <Compile Include="Traits\BotModules\HarvesterBotModule.cs" />
-    <Compile Include="AI\AISupportPowerManager.cs" />
+    <Compile Include="Traits\BotModules\\SupportPowerBotModule.cs" />
```
Comment (pchote) [inline OpenRA.Mods.Common/OpenRA.Mods.Common.csproj]: Double backslash in this path.

File: OpenRA.Mods.Common/Traits/BotModules/SupportPowerBotModule.cs [hunk 2]
```diff
@@ -13,38 +13,71 @@
...
 
-		public void TryToUseSupportPower(Actor self)
+		void ITick.Tick(Actor self)
 		{
+			if (IsTraitDisabled)
+				return;
+
 			if (supportPowerManager == null)
+			{
+				AIUtils.BotDebug("AI: {0} can't use support powers due to lack of SupportPowerManager.", player.PlayerName);
```
Comment (pchote) [inline OpenRA.Mods.Common/Traits/BotModules/SupportPowerBotModule.cs]: Logging this 25 times per second for the entire game isn't a good idea.
We can make this trait `Requires<SupportPowerManagerInfo>` and drop this check completely.

File: OpenRA.Mods.Common/UpdateRules/Rules/20180923/AddSupportPowerBotModule.cs [hunk 3]
```diff
@@ -0,0 +1,95 @@
...
+ */
+#endregion
+
+using System;
+using System.Collections.Generic;
+using System.Linq;
+
+namespace OpenRA.Mods.Common.UpdateRules.Rules
+{
+	public class AddSupportPowerBotModule : UpdateRule
```
Comment (pchote) [inline OpenRA.Mods.Common/UpdateRules/Rules/20180923/AddSupportPowerBotModule.cs]: How ugly would it be to combine these all into a single `ExtractHackyAIModules` rule?
It seems silly to have N different classes that are half duplicates of eachother.

**Comment by reaperrr** [comment]
Rebased.

**Comment by reaperrr** [comment]
Updated, renamed `SupportPowerDecisions` to just `Decisions` as suggested by @Phrohdoh.

**Comment by pchote** [comment]
Adding dependencies tag because of the issue I raised in #15750.

**Comment by reaperrr** [comment]
Rebased.

File: OpenRA.Mods.Common/AI/HackyAI.cs [hunk 4]
```diff
@@ -245,6 +229,7 @@ static object LoadDecisions(MiniYaml yaml)
 
 	public sealed class HackyAI : ITick, IBot, INotifyDamage
 	{
+		// DEPRECATED: Modules should define a MersenneTwister themselves or use World.LocalRandom.
```
Comment (pchote) [inline OpenRA.Mods.Common/AI/HackyAI.cs]: I don't see a usecase for modules to define their own RNG, so better to remove this IMO.

File: OpenRA.Mods.Common/Traits/BotModules/SupportPowerBotModule.cs [hunk 5]
```diff
@@ -10,37 +10,60 @@
...
 using OpenRA.Traits;
 
-namespace OpenRA.Mods.Common.AI
+namespace OpenRA.Mods.Common.Traits
 {
-	class AISupportPowerManager
+	[Desc("Manages bot support power handling.")]
+	public class SupportPowerBotModuleInfo : ConditionalTraitInfo, Requires<SupportPowerManagerInfo>
+	{
+		// TODO Update OpenRA.Utility/Command.cs#L300 to first handle lists and also read nested ones
```
Comment (pchote) [inline OpenRA.Mods.Common/Traits/BotModules/SupportPowerBotModule.cs]: This file doesn't exist, and i'm not sure whether this limitation still applies!

**Comment by pchote** [comment]
Not directly related to the changes in this PR, but the update rule crashes when I try to apply it:

```
ExtractHackyAIModules: Split HackyAI logic handling to BotModules
   Updating mod... COMPLETE
   Updating system maps... FAILED

   The automated changes for this rule were not applied because of an error.
   After the issue reported below is resolved you should run the updater
   with SOURCE set to ExtractHackyAIModules to retry these changes

   The map that caused the error was:
     /Users/paul/src/OpenRA/mods/ra/maps/fort-lonestar

   The exception reported was:
     System.NullReferenceException: Object reference not set to an instance of an object
       at OpenRA.Mods.Common.UpdateRules.UpdateExtensions.NodeValue[T] (OpenRA.MiniYamlNode node) [0x00001] in /Users/paul/src/OpenRA/OpenRA.Mods.Common//UpdateRules/UpdateUtils.cs:253 
       at OpenRA.Mods.Common.UpdateRules.Rules.ExtractHackyAIModules+<UpdateActorNode>c__Iterator1.MoveNext () [0x000a8] in /Users/paul/src/OpenRA/OpenRA.Mods.Common//UpdateRules/Rules/20180923/ExtractHackyAIModules.cs:70 
       at OpenRA.Mods.Common.UpdateRules.UpdateUtils+<ApplyTopLevelTransform>c__Iterator3.MoveNext () [0x00127] in /Users/paul/src/OpenRA/OpenRA.Mods.Common//UpdateRules/UpdateUtils.cs:220 
       at System.Collections.Generic.List`1[T].AddEnumerable (System.Collections.Generic.IEnumerable`1[T] enumerable) [0x00059] in <016f22b59bd64951b06e38b9f5a0506c>:0 
       at System.Collections.Generic.List`1[T].InsertRange (System.Int32 index, System.Collections.Generic.IEnumerable`1[T] collection) [0x000f4] in <016f22b59bd64951b06e38b9f5a0506c>:0 
       at System.Collections.Generic.List`1[T].AddRange (System.Collections.Generic.IEnumerable`1[T] collection) [0x00000] in <016f22b59bd64951b06e38b9f5a0506c>:0 
       at OpenRA.Mods.Common.UpdateRules.UpdateUtils.UpdateMap (OpenRA.ModData modData, OpenRA.FileSystem.IReadWritePackage mapPackage, OpenRA.Mods.Common.UpdateRules.UpdateRule rule, System.Collections.Generic.List`1[System.Tuple`3[OpenRA.FileSystem.IReadWritePackage,System.String,System.Collections.Generic.List`1[OpenRA.MiniYamlNode]]]& files, System.Collections.Generic.HashSet`1[T] externalFilenames) [0x000c7] in /Users/paul/src/OpenRA/OpenRA.Mods.Common//UpdateRules/UpdateUtils.cs:108 
       at OpenRA.Mods.Common.UtilityCommands.UpdateModCommand.ApplyRules (OpenRA.ModData modData, System.Collections.Generic.IEnumerable`1[T] rules, System.Boolean skipMaps) [0x00186] in /Users/paul/src/OpenRA/OpenRA.Mods.Common//UtilityCommands/UpdateModCommand.cs:206 
```

The rule will need some tweaks to fail cleanly, ideally in its own PR.

**Review by pchote — DISMISSED** [review]
LGTM, just two minor nits:

**Comment by reaperrr** [comment]
Updated.

**Review by obrakmann — APPROVED** [review]

**Comment by obrakmann** [comment]
[Changelog](https://github.com/OpenRA/OpenRA/wiki/Changelog-(bleed)/_compare/e906bd7c1d6777e4a67036e822df90217d40b2ad)

**Pull request merged**